text/0089-jetton-wallet-discovery.md

```diff
@@ -0,0 +1,71 @@
+- **TEP**: [89](https://github.com/ton-blockchain/TEPs/pull/89)
+- **title**: Discoverable Jettons Wallets
+- **status**: Draft
+- **type**: Contract Interface
+- **authors**: [EmelyanenkoK](https://github.com/EmelyanenkoK)
+- **created**: 08.09.2022 
+- **replaces**: -
+- **replaced by**: -
+
+# Summary
+
+This proposal suggest to extend standard Jetton master by adding mandatory onchain `get_wallet_address` handler.
+
+# Motivation
+
+Some application may want to be able to discover their or other contract wallets for some specific Jetton Master. For instance some contract may want to obtain and store it's jetton wallet for some Jetton to handle transfer notifications from it in specific way.
+
+# Guide
+
+Upon receiving `get_wallet_address` message with address in question, Jetton Master should response with message containing address.
+
+# Specification
+
+## New Jetton Master contracts
+Jettom Master should handle message
+
+`get_wallet_address#418cbb4e query_id:uint64 owner_address:MsgAddress include_address:Bool = InternalMsgBody;`
+
+with TON amount higher than `5000 gas-units + msg_forward_prices.lump_price + msg_forward_prices.cell_price` = 0.0061 TON for current basechain settings (if amount is less than that it is not possible to send response) attached
+
+and either throw an exception if amount of incoming value is not enough to calculate wallet address or
+response with message (sent with mode 64)
+
+`report_wallet_address#ad30f94a query_id:uint64 wallet_address:MsgAddress owner_address:(Maybe ^MsgAddress) = InternalMsgBody;`
```

Comment:
Please consider naming report_wallet_address return_wallet_address, because report has the context of marking an entity as malicious etc.

Reply:
other potential terms instead of "report" are "return" / "query" / "get"

Reply:
Currently we mimic names used in NFT for `get_static_data`.
I don't like return/query naming because I think it will looks a little bit counterintuitive to implement "return_wallet_address" method on recipient side to actually handle the response.
What about a `take_static_data`? Alternatively we can just use `get_static_data` op-code with high-bit set to true (to emphasize it is response), however we doesn't have TL-B notation for that.
Anyway, I don't think that this issue is important and any decision will work ok.

Reply:
As I can see, the best way to solve the problem is to use a synonym of the word "report". 

I suggest using the word "inform". It is neutral and doesn't have a negative meaning. (e.g. inform_wallet_address, inform_static_data)

PS: imho, we should decide faster so people can start to use an up-to-date jetton standard.

Reply:
We realised that op should represent the **operation that receiver should execute**.

With that in mind the best way to name our ops is: provide_wallet_address and take_wallet_address.

- Smart-contract sends op:provide_wallet_address.
- Discoverable minter receives op:provide_wallet_address and sends op:take_wallet_address.
- Smart-contract receives op:take_wallet_address and handle the message.